File: files/arc_openmp/fortran/race.f90
program hello_world

    ! Load the OpenMP functions library
    use omp_lib

    implicit none

    integer :: i, best_cost, tid, my_cost
    integer, parameter :: N = 20

    ! Seed random number
    call random_seed(put=(/0/))
    best_cost=huge(best_cost)

    !  Fork a team of threads, with private versions of the declared variables.
    !$OMP PARALLEL SHARED(best_cost) PRIVATE(tid, my_cost)

    !$OMP DO
    do i = 1, N
      ! Note check thread-safety of random generator.
      call random_number(my_cost)
      my_cost = floor(my_cost*100)
      print *, "tid ", tid, " - ", tid, my_cost
      !$OMP OMP CRITICAL
      if (best_cost > my_cost) then
        print *, "tid ", tid, " says that ", my_cost, " is lower than ", best_cost
        best_cost = my_cost
      end if
    end do
    !$OMP END DO NOWAIT

    ! Disband all but the master thread
    !$OMP END PARALLEL

    print *, "Best cost ", best_cost
    
end program hello_world

File: files/arc_openmp/c/race.c
#include<omp.h>
#include<stdlib.h>

int main()
{
    srand (0);
    int i;
    int N=20;
    int my_rand[N];
    int best_cost=RAND_MAX;

    for (i=0; i<N; i++)
    {
      my_rand[i] = rand();
    }

    #pragma omp parallel shared(best_cost)
    {
        #pragma omp for nowait
        for (i=0; i<N; i++)
        {
            int tid = omp_get_thread_num();
            int my_cost;
            // Be careful using rand in OpenMP parallel regions.
            my_cost = my_rand[i] % 100;
            printf("tid %i -  %i\n",tid,my_cost);
            //#pragma omp critical
            if(best_cost > my_cost)
            {
                printf("tid %i says that %i is lower than %i \n",tid,my_cost,best_cost);
                best_cost = my_cost;
            }
        }
     }
     printf("Best cost %d\n",best_cost);
}

